### PR TITLE
Add RADIO Vision Encoder Support to vLLM

### PR DESCRIPTION
**Pull request by danielafrimi**
This PR implements support for the [C-RADIO](https://huggingface.co/nvidia/C-RADIO) (Retrieval-Augmented Dual Instruction Optimization) vision encoder in vLLM, enabling its use with multimodal models like Nano Nemotron VL.

### Changes

**New Radio Model Implementation** (`vllm/model_executor/models/radio.py`)
- `RadioInternVisionModel`: Core vision model using **InternVision** encoder architecture

**Integration Updates** (`vllm/model_executor/models/nano_nemotron_vl.py`)
- Updated Nano Nemotron VL to use the new RadioModel for vision processing

**Testing** (`tests/models/multimodal/pooling/test_radio.py`)
- Comprehensive tests for RADIO model with `nvidia/C-RADIOv2-H`
- Validates output consistency between HuggingFace and vLLM implementations

### Technical Notes

**Hardcoded Values**: The implementation preserves hardcoded values from the original `timm` package implementation, including OpenAI CLIP normalization constants and predefined ViT model dimensions, ensuring compatibility and reproducibility.

**Configuration**: Create new configuration approach to instantiate the Radio model based on InterVision model architecture, with dynamic parameter mapping for different ViT variants.

**Weight Loading**: Custom weight loader handles mapping between HuggingFace and vLLM parameter names, supporting models with `radio_model.` prefix while skipping unused parameters.

### CODE REVIEW

File: tests/models/multimodal/pooling/test_radio.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+from vllm.model_executor.models.radio import RadioModel
+from vllm.utils import STR_DTYPE_TO_TORCH_DTYPE
+
+from ....conftest import ImageTestAssets
+
+# we use snapshot_download to prevent conflicts between
+# dynamic_module and trust_remote_code for hf_runner
+DOWNLOAD_PATTERN = ["*.json", "*.py", "*.safetensors", "*.txt", "*.model"]
+
+VIT_DIMS = {
```
Comment (danielafrimi) [inline tests/models/multimodal/pooling/test_radio.py]: The` config.json` of C-RADIO includes the model attr, and based on this name they infer attributes such as hidden_size, num_layers, num_heads, and intermediate_size. 

In my test/modeling I created a dictionary for this mapping, but I’m not sure where the best place would be to keep it.

[ref](https://huggingface.co/nvidia/C-RADIO/blob/main/config.json#L106) 

WDYT?

File: tests/models/multimodal/pooling/test_radio.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            return_tensors='pt').pixel_values.to(torch_dtype)[:, :, :, :640]
+        for images in images
+    ]
+
+    config = AutoConfig.from_pretrained(model_id, trust_remote_code=True)
+
+    hidden_size, num_layers, num_heads, intermediate_size = (
+        get_args_from_model_type(config.args["model"]))
+
+    config.num_hidden_layers = num_layers
```
Comment (danielafrimi) [inline tests/models/multimodal/pooling/test_radio.py]: These hardcoded values correspond to the ones implemented in the `timm` package. 

I extracted them from the original model implementation and set them as fixed values. 
For the` nano_nemotron_vl `model, we could probably add them into the `config.json`, but I’m not sure how we should handle cases where someone might want to use the model with different values. 

WDYT?

File: tests/models/multimodal/pooling/test_radio.py (hunk 1 above)
Comment (DarkLight1337) [reply tests/models/multimodal/pooling/test_radio.py]: Just put it in the modeling file. And you can import the modeling file from this file inside the test function (to avoid early CUDA initialization)

File: vllm/model_executor/models/radio.py [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+
+
+to_1tuple = _ntuple(1)
+to_2tuple = _ntuple(2)
+to_3tuple = _ntuple(3)
+to_4tuple = _ntuple(4)
+to_ntuple = _ntuple
+
+
+class InputConditioner(nn.Module):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/radio.py:47]: QQ: Why is the input conditioner applied both in `NanoNemotronVLProcessor` and here?

File: tests/models/multimodal/pooling/test_radio.py (hunk 1 above)
Comment (danielafrimi) [reply tests/models/multimodal/pooling/test_radio.py]: done

File: tests/models/multimodal/pooling/test_radio.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    images = [asset.pil_image for asset in image_assets]
+    # Input resolution must be a multiple of `self.min_resolution_step`.
+    # Using `self.get_nearest_supported_resolution`, for assets 432x642 the
+    # nearest supported resolution is 432x640.
+    pixel_values = [
+        img_processor(
+            images,
+            return_tensors='pt').pixel_values.to(torch_dtype)[:, :, :, :640]
+        for images in images
+    ]
```
[the remark this answers is not in the document]
Comment (danielafrimi) [reply tests/models/multimodal/pooling/test_radio.py]: fixed naming

File: vllm/model_executor/models/radio.py (hunk 3 above)
Comment (danielafrimi) [reply vllm/model_executor/models/radio.py:47]: thanks, removed it and set the input conditioner in model only.

File: tests/models/multimodal/pooling/test_radio.py (hunk 2 above)
Comment (danielafrimi) [reply tests/models/multimodal/pooling/test_radio.py]: @DarkLight1337 still wondering about this

**Comment by danielafrimi** [comment]
@DarkLight1337 Fixed the comments.

File: tests/models/multimodal/pooling/test_radio.py (hunk 2 above)
Comment (DarkLight1337) [reply tests/models/multimodal/pooling/test_radio.py]: I think we can create a custom config under `vllm/transformers_utils` for this

Comment (danielafrimi) [reply tests/models/multimodal/pooling/test_radio.py]: done

File: vllm/model_executor/models/radio.py [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+
+        token = self.token.unsqueeze(0).expand(x.shape[0], -1, -1)
+        x = torch.cat([
+            token,
+            x,
+        ], dim=1)
+
+        return x
+
+    def no_weight_decay(self):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/radio.py]: Unused method

File: vllm/model_executor/models/radio.py [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+
+    @property
+    def num_registers(self):
+        return self.cls_token.num_registers
+
+    @property
+    def num_skip(self):
+        return self.num_cls_tokens + self.num_registers
+
+    def no_weight_decay(self):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/radio.py]: Unused method

File: vllm/model_executor/models/radio.py [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+            scale = ndim**-0.5
+            self.token = nn.Parameter(
+                torch.randn(num_tokens + self.num_registers, ndim) * scale)
+
+        else:
+            self.token = None
+
+        self.num_patches = self.num_tokens + self.num_registers
+
+    def disable(self):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/radio.py]: Unused method

Comment (danielafrimi) [reply vllm/model_executor/models/radio.py]: removed

File: vllm/model_executor/models/radio.py (hunk 6 above)
Comment (danielafrimi) [reply vllm/model_executor/models/radio.py]: removed

File: vllm/model_executor/models/radio.py (hunk 5 above)
Comment (danielafrimi) [reply vllm/model_executor/models/radio.py]: removed

File: vllm/model_executor/models/radio.py [hunk 8]
```diff
@@ -0,0 +1,582 @@
...
+    def forward(
+        self,
+        pixel_values: Optional[torch.Tensor] = None,
+        pixel_embeds: Optional[torch.Tensor] = None,
+    ) -> torch.FloatTensor:
+        x = self.input_conditioner(pixel_values)
+        y = self.model(x)
+        return self._extract_final(y)
+
+    def make_preprocessor_external(
```
Comment (DarkLight1337) [inline vllm/model_executor/models/radio.py]: This method can also be removed

Comment (danielafrimi) [reply vllm/model_executor/models/radio.py]: removed

**Review by DarkLight1337 — APPROVED** [review]
LGTM now, thanks
